bpf/tc/attach.go

```diff
@@ -130,7 +164,19 @@ func (ap AttachPoint) AttachProgram() error {
 				Stderr:   string(err.Stderr),
 			}
 		}
-		return errors.Wrap(err, "failed to attach TC program")
+		return errors.WithMessage(err, "failed to attach TC program")
+	}
+
+	// Success: clean up the old programs.
+	for _, p := range progsToClean {
+		log.WithField("prog", p).Debug("Cleaning up old calico program")
+		tcCmd := exec.Command("tc", "filter", "del", "dev", ap.Iface, string(ap.Hook), "pref", p.pref, "handle", p.handle, "bpf")
+		err := tcCmd.Run()
+		if err != nil {
+			// TODO: filter error to avoid spam if interface is gone.
+			log.WithError(err).WithField("prog", p).Warn("Failed to clean up old calico program.")
+			return errors.WithMessage(err, "failed to clean up old program")
```

Comment:
Would it be better to try to clean up the whole of progsToClean, before returning an error?

Reply:
Yes, good catch.